docs/decisions/0006-tanstack-react-query.rst

```diff
@@ -0,0 +1,71 @@
+
+
+We will adopt query key factories to manage the implementation of query keys such that cache invalidation for independent features (e.g., Learner Credit Management) can be managed with adequate granularity. For example:
+
+::
+
+  // Query Key Factory
+  export const learnerCreditManagementQueryKeys = {
+    all: ['learner-credit-management'],
+    budgets: () => [...learnerCreditManagementQueryKeys.all, 'budgets'],
+    budget: (budgetId) => [...learnerCreditManagementQueryKeys.all, 'budget', budgetId],
+    budgetActivity: (budgetId) => [...learnerCreditManagementQueryKeys.budget(budgetId), 'activity'],
+    budgetActivityOverview: (budgetId) => [...learnerCreditManagementQueryKeys.budgetActivity(budgetId), 'overview'],
+  };
+
+By having a query key factory as suggested above, contributors may have a structured way to manage query keys for a given feature. This approach enabled granular control over cache invalidation, query prefetching, etc. Using the above example, one could invalidate the query cache for the entire ``['learner-credit-management']`` feature or opt to only invalidate the query cache for specific individual queries or even groups of queries:
+
+::
+
+  // Remove everything related to the learner credit management feature
+  queryClient.removeQueries({
+    queryKey: learnerCreditManagementQueryKeys.all,
+  })
+
+  // Invalidate all queries supporting the budget detail page route
+  queryClient.invalidateQueries({
+    queryKey: learnerCreditManagementQueryKeys.budget(budgetId),
+  })
+
+  // Invalidate the budget detail page route's activity tab's overview query
+  queryClient.invalidateQueries({
+    queryKey: learnerCreditManagementQueryKeys.budgetActivityOverview(budgetId),
+  })
+
+The recommendation for new asynchronous network calls moving forward is to use ``@tanstack/react-query``. Additionally, it's recommended to incrementally migrate existing network calls to use ``@tanstack/react-query``. This will allow us to avoid a large refactoring effort and instead migrate to the library over time as we touch existing network calls.
+
+Consequences
+************
+
+The entire application is wrapped within ``QueryClientProvider``, which contains a default configuration to at least extend the ``staleTime`` to be 20 seconds. This change in default behavior is to enable the use of ``@tanstack/react-query`` as a state management. Instead of queries becoming instantly stale after success (i.e., ``staleTime: 0``) where network calls would be re-fetched on all window refocuses and component mounts, etc., having a ``staleTime`` of 20 seconds will keep the data fresh for 20 seconds before the query becomes stale, preventing unnecessary API calls (e.g., when calling duplicate ``useQuery`` hooks in the same rendering lifecycle).
```

Comment:
nit: "as a state management _tool_"? or "state manager"

Reply:
I think either could work? I'm using "state manager" based on https://tkdodo.eu/blog/react-query-as-a-state-manager.

Reply:
Fixed the typo. Good catch!